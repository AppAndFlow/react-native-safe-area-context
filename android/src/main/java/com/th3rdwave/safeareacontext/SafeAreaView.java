package com.th3rdwave.safeareacontext;

import static com.facebook.react.uimanager.UIManagerHelper.getReactContext;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.WindowInsets;

import androidx.annotation.Nullable;

import com.facebook.react.bridge.ReactContext;
import com.facebook.react.views.view.ReactViewGroup;
import com.facebook.react.uimanager.UIManagerModule;

import java.util.EnumSet;

@SuppressLint("ViewConstructor")
public class SafeAreaView extends ReactViewGroup implements ViewTreeObserver.OnGlobalLayoutListener {
  private @Nullable EdgeInsets mInsets;
  private @Nullable EnumSet<SafeAreaViewEdges> mEdges;

  public SafeAreaView(Context context) {
    super(context);
  }

  private void updateInsets() {
    if (mInsets != null) {
      EnumSet<SafeAreaViewEdges> edges = mEdges != null
              ? mEdges
              : EnumSet.allOf(SafeAreaViewEdges.class);

      SafeAreaViewLocalData localData = new SafeAreaViewLocalData(mInsets, edges);

      ReactContext reactContext = getReactContext(this);
      UIManagerModule uiManager = reactContext.getNativeModule(UIManagerModule.class);
      if (uiManager != null) {
        uiManager.setViewLocalData(getId(), localData);
      }
    }
  }

  public void setEdges(EnumSet<SafeAreaViewEdges> edges) {
    mEdges = edges;
    updateInsets();
  }

  private void maybeUpdateInsets() {
    EdgeInsets edgeInsets = SafeAreaUtils.getSafeAreaInsets(getRootView(), this);
    if (edgeInsets != null && (mInsets == null || !mInsets.equalsToEdgeInsets(edgeInsets))) {
      mInsets = edgeInsets;
      updateInsets();
    }
  }

  @Override
  protected void onAttachedToWindow() {
    super.onAttachedToWindow();

    getViewTreeObserver().addOnGlobalLayoutListener(this);
    maybeUpdateInsets();
  }

  @Override
  protected void onDetachedFromWindow() {
    super.onDetachedFromWindow();

    getViewTreeObserver().removeOnGlobalLayoutListener(this);
  }

  @Override
  public void onGlobalLayout() {
    maybeUpdateInsets();
  }

  @Override
  protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
    maybeUpdateInsets();
    super.onLayout(changed, left, top, right, bottom);
  }
}
