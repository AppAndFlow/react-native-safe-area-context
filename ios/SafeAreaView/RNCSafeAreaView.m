#import "RNCSafeAreaView.h"

#import <React/RCTBridge.h>
#import <React/RCTUIManager.h>

#import "RNCSafeAreaViewLocalData.h"
#import "RNCSafeAreaViewMode.h"
#import "RNCSafeAreaViewEdges.h"
#import "RCTView+SafeAreaCompat.h"

@implementation RNCSafeAreaView {
  __weak RCTBridge *_bridge;
  UIEdgeInsets _currentSafeAreaInsets;
  RNCSafeAreaViewMode _mode;
  RNCSafeAreaViewEdges _edges;
}

- (instancetype)initWithBridge:(RCTBridge *)bridge
{
  if (self = [super initWithFrame:CGRectZero]) {
    _bridge = bridge;
    // Defaults
    _emulateUnlessSupported = YES;
    _mode = RNCSafeAreaViewModePadding;
    _edges = RNCSafeAreaViewEdgesAll;
  }

  return self;
}

RCT_NOT_IMPLEMENTED(-(instancetype)initWithCoder : (NSCoder *)decoder)
RCT_NOT_IMPLEMENTED(-(instancetype)initWithFrame : (CGRect)frame)

- (NSString *)description
{
  NSString *superDescription = [super description];

  // Cutting the last `>` character.
  if (superDescription.length > 0 && [superDescription characterAtIndex:superDescription.length - 1] == '>') {
    superDescription = [superDescription substringToIndex:superDescription.length - 1];
  }

  return [NSString stringWithFormat:@"%@; RNCSafeAreaInsets = %@; appliedRNCSafeAreaInsets = %@>",
          superDescription,
          NSStringFromUIEdgeInsets([self viewControllerOrEmulateSafeAreaInsets:self.emulateUnlessSupported]),
          NSStringFromUIEdgeInsets(_currentSafeAreaInsets)];
}

- (void)safeAreaInsetsDidChange
{
  [super safeAreaInsetsDidChange];
  [self invalidateSafeAreaInsets];
}

- (void)layoutSubviews
{
  [super layoutSubviews];

  if (!self.nativeSafeAreaSupport && self.emulateUnlessSupported) {
    [self invalidateSafeAreaInsets];
  }
}

- (void)invalidateSafeAreaInsets
{
  UIEdgeInsets safeAreaInsets = [self viewControllerOrEmulateSafeAreaInsets:self.emulateUnlessSupported];

  if (UIEdgeInsetsEqualToEdgeInsetsWithThreshold(safeAreaInsets, _currentSafeAreaInsets, 1.0 / RCTScreenScale())) {
    return;
  }

  _currentSafeAreaInsets = safeAreaInsets;
  [self updateLocalData];
}

- (void)updateLocalData
{
  RNCSafeAreaViewLocalData *localData = [[RNCSafeAreaViewLocalData alloc] initWithInsets:_currentSafeAreaInsets
                                                                                    mode:_mode
                                                                                   edges:_edges];
  [_bridge.uiManager setLocalData:localData forView:self];
}

- (void)setEmulateUnlessSupported:(BOOL)emulateUnlessSupported
{
  if (_emulateUnlessSupported == emulateUnlessSupported) {
    return;
  }

  _emulateUnlessSupported = emulateUnlessSupported;

  if ([self nativeSafeAreaSupport]) {
    return;
  }

  [self invalidateSafeAreaInsets];
}

- (void)setMode:(RNCSafeAreaViewMode)mode {
    _mode = mode;
    [self updateLocalData];
}

- (void)setEdges:(RNCSafeAreaViewEdges)edges {
  _edges = edges;
  [self updateLocalData];
}

@end
