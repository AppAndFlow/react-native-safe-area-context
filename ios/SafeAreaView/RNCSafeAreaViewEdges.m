#import <React/RCTConvert.h>
#import "RNCSafeAreaViewEdges.h"

@implementation RCTConvert (RNCSafeAreaView)

RCT_ENUM_CONVERTER(RNCSafeAreaViewEdges, (@{
  @"none": @(RNCSafeAreaViewEdgesNone),
  
  @"top": @(RNCSafeAreaViewEdgesTop),
  @"right": @(RNCSafeAreaViewEdgesRight),
  @"bottom": @(RNCSafeAreaViewEdgesBottom),
  @"left": @(RNCSafeAreaViewEdgesLeft),
  
  @"horizontal": @(RNCSafeAreaViewEdgesRight | RNCSafeAreaViewEdgesLeft),
  @"vertical": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesBottom),
  @"top-right": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesRight),
  @"top-left": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesLeft),
  @"bottom-right": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesRight),
  @"bottom-left": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesLeft),
  
  @"not-top": @(RNCSafeAreaViewEdgesRight | RNCSafeAreaViewEdgesBottom | RNCSafeAreaViewEdgesLeft),
  @"not-right": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesBottom | RNCSafeAreaViewEdgesLeft),
  @"not-bottom": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesRight | RNCSafeAreaViewEdgesLeft),
  @"not-left": @(RNCSafeAreaViewEdgesTop | RNCSafeAreaViewEdgesRight | RNCSafeAreaViewEdgesBottom),
  
  @"all": @(RNCSafeAreaViewEdgesAll),
}), RNCSafeAreaViewEdgesNone, integerValue);

@end
