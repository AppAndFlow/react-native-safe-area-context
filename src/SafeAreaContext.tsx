import * as React from 'react';
import { Dimensions, StyleSheet, ViewProps } from 'react-native';
import { NativeSafeAreaProvider } from './NativeSafeAreaProvider';
import type {
  EdgeInsets,
  InsetChangedEvent,
  Metrics,
  InitialMetricsState,
  Rect,
} from './SafeArea.types';

const isDev = process.env.NODE_ENV !== 'production';

export const SafeAreaInsetsContext = React.createContext<EdgeInsets | null>(
  null,
);
if (isDev) {
  SafeAreaInsetsContext.displayName = 'SafeAreaInsetsContext';
}

export const SafeAreaFrameContext = React.createContext<Rect | null>(null);
if (isDev) {
  SafeAreaFrameContext.displayName = 'SafeAreaFrameContext';
}

export const SafeAreaGetLatestWindowMetricsContext = React.createContext<
  (() => InitialMetricsState | Metrics) | null
>(null);
if (isDev) {
  SafeAreaGetLatestWindowMetricsContext.displayName =
    'SafeAreaGetLatestWindowMetricsContext';
}

export const SafeAreaAddListenerContext = React.createContext<
  ((callback: (metrics: Metrics) => void) => () => void) | null
>(null);

export interface SafeAreaProviderProps extends ViewProps {
  children?: React.ReactNode;
  initialMetrics?: Metrics | null;
  /**
   * @deprecated
   */
  initialSafeAreaInsets?: EdgeInsets | null;
}

export function SafeAreaProvider({
  children,
  initialMetrics,
  initialSafeAreaInsets,
  style,
  ...others
}: SafeAreaProviderProps) {
  const listeners = React.useRef<((metrics: Metrics) => void)[]>([]);

  const parentInsets = useParentSafeAreaInsets();
  const parentFrame = useParentSafeAreaFrame();

  const [metrics, setMetrics] = React.useState<Metrics | InitialMetricsState>({
    frame: initialMetrics?.frame ??
      parentFrame ?? {
        // Backwards compat so we render anyway if we don't have frame.
        x: 0,
        y: 0,
        width: Dimensions.get('window').width,
        height: Dimensions.get('window').height,
      },

    insets:
      initialMetrics?.insets ?? initialSafeAreaInsets ?? parentInsets ?? null,
  });

  const latestMetrics = React.useRef(metrics);

  const onInsetsChange = React.useCallback(
    (event: InsetChangedEvent) => {
      const {
        nativeEvent: { frame: nextFrame, insets: nextInsets },
      } = event;

      setMetrics((currentMetrics) => {
        const { frame, insets } = currentMetrics;

        const didFrameChange =
          nextFrame &&
          (nextFrame.height !== frame.height ||
            nextFrame.width !== frame.width ||
            nextFrame.x !== frame.x ||
            nextFrame.y !== frame.y);

        const didInsetsChange =
          !insets ||
          nextInsets.bottom !== insets.bottom ||
          nextInsets.left !== insets.left ||
          nextInsets.right !== insets.right ||
          nextInsets.top !== insets.top;

        if (didFrameChange || didInsetsChange) {
          const nextMetrics = {
            frame: didFrameChange ? nextFrame : frame,
            insets: didInsetsChange ? nextInsets : insets,
          };

          latestMetrics.current = nextMetrics;

          listeners.current.forEach((listener) => listener(nextMetrics));

          return nextMetrics;
        } else {
          return currentMetrics;
        }
      });
    },
    [setMetrics],
  );

  const getLatestWindowMetrics = React.useCallback(() => {
    return latestMetrics.current;
  }, [latestMetrics]);

  const addListener = React.useCallback(
    (listener: (metrics: Metrics) => void) => {
      listeners.current.push(listener);

      const unsubscribe = () => {
        const index = listeners.current.indexOf(listener);
        if (index > -1) {
          listeners.current.splice(index, 1);
        }
      };

      return unsubscribe;
    },
    [listeners],
  );

  return (
    <NativeSafeAreaProvider
      style={StyleSheet.compose(styles.fill, style)}
      onInsetsChange={onInsetsChange}
      {...others}
    >
      {metrics.insets != null ? (
        <SafeAreaFrameContext.Provider value={metrics.frame}>
          <SafeAreaInsetsContext.Provider value={metrics.insets}>
            <SafeAreaGetLatestWindowMetricsContext.Provider
              value={getLatestWindowMetrics}
            >
              <SafeAreaAddListenerContext.Provider value={addListener}>
                {children}
              </SafeAreaAddListenerContext.Provider>
            </SafeAreaGetLatestWindowMetricsContext.Provider>
          </SafeAreaInsetsContext.Provider>
        </SafeAreaFrameContext.Provider>
      ) : null}
    </NativeSafeAreaProvider>
  );
}

const styles = StyleSheet.create({
  fill: { flex: 1 },
});

function useParentSafeAreaInsets(): EdgeInsets | null {
  return React.useContext(SafeAreaInsetsContext);
}

function useParentSafeAreaFrame(): Rect | null {
  return React.useContext(SafeAreaFrameContext);
}

const NO_INSETS_ERROR =
  'No safe area value available. Make sure you are rendering `<SafeAreaProvider>` at the top of your app.';

export function useSafeAreaInsets(): EdgeInsets {
  const insets = React.useContext(SafeAreaInsetsContext);
  if (insets == null) {
    throw new Error(NO_INSETS_ERROR);
  }
  return insets;
}

export function useSafeAreaFrame(): Rect {
  const frame = React.useContext(SafeAreaFrameContext);
  if (frame == null) {
    throw new Error(NO_INSETS_ERROR);
  }
  return frame;
}

export function useSafeAreaGetLatestWindowMetrics() {
  const getLatestWindowMetrics = React.useContext(
    SafeAreaGetLatestWindowMetricsContext,
  );
  if (getLatestWindowMetrics == null) {
    throw new Error(NO_INSETS_ERROR);
  }
  return getLatestWindowMetrics;
}

export function useSafeAreaAddListener() {
  const addListener = React.useContext(SafeAreaAddListenerContext);
  if (addListener == null) {
    throw new Error(NO_INSETS_ERROR);
  }
  return addListener;
}

export type WithSafeAreaInsetsProps = {
  insets: EdgeInsets;
};

export function withSafeAreaInsets<T>(
  WrappedComponent: React.ComponentType<T & WithSafeAreaInsetsProps>,
): React.ForwardRefExoticComponent<
  React.PropsWithoutRef<T> & React.RefAttributes<unknown>
> {
  return React.forwardRef((props: T, ref: React.Ref<unknown>) => {
    const insets = useSafeAreaInsets();
    return <WrappedComponent {...props} insets={insets} ref={ref} />;
  });
}

/**
 * @deprecated
 */
export function useSafeArea(): EdgeInsets {
  return useSafeAreaInsets();
}

/**
 * @deprecated
 */
export const SafeAreaConsumer = SafeAreaInsetsContext.Consumer;

/**
 * @deprecated
 */
export const SafeAreaContext = SafeAreaInsetsContext;

function useLatest<T>(value: T) {
  const ref = React.useRef(value);
  ref.current = value;

  return ref;
}

export default useLatest;
