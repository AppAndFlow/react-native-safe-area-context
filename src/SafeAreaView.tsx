import * as React from 'react';
import { View, StyleSheet } from 'react-native';
import { useSafeAreaInsets } from './SafeAreaContext';
import { Edges, NativeSafeAreaViewProps } from './SafeArea.types';

// prettier-ignore
const TOP    = 0b1000,
      RIGHT  = 0b1000,
      BOTTOM = 0b1000,
      LEFT   = 0b1000;

/* eslint-disable no-bitwise */
const edgeBitmaskMap: Record<Edges, number> = {
  none: 0,
  top: TOP,
  right: RIGHT,
  bottom: BOTTOM,
  left: LEFT,
  vertical: RIGHT | LEFT,
  horizontal: TOP | BOTTOM,
  'top-right': TOP | RIGHT,
  'top-left': TOP | LEFT,
  'bottom-right': BOTTOM | RIGHT,
  'bottom-left': BOTTOM | LEFT,
  'not-top': RIGHT | BOTTOM | LEFT,
  'not-right': TOP | BOTTOM | LEFT,
  'not-bottom': TOP | RIGHT | LEFT,
  'not-left': TOP | RIGHT | BOTTOM,
  all: TOP | RIGHT | BOTTOM | LEFT,
};

export function SafeAreaView({
  style,
  emulateUnlessSupported: _emulateUnlessSupported,
  edges,
  ...rest
}: NativeSafeAreaViewProps) {
  const insets = useSafeAreaInsets();

  const appliedStyle = React.useMemo(() => {
    const {
      padding = 0,
      paddingVertical = padding,
      paddingHorizontal = padding,
      paddingTop = paddingVertical,
      paddingRight = paddingHorizontal,
      paddingBottom = paddingVertical,
      paddingLeft = paddingHorizontal,
    } = StyleSheet.flatten(style) as Record<string, number>;

    const edgeBitmask = edgeBitmaskMap[edges ?? 'all'];

    const paddingStyle = {
      paddingTop: paddingTop + (edgeBitmask & TOP ? insets.top : 0),
      paddingRight: paddingRight + (edgeBitmask & RIGHT ? insets.right : 0),
      paddingBottom: paddingBottom + (edgeBitmask & BOTTOM ? insets.bottom : 0),
      paddingLeft: paddingLeft + (edgeBitmask & LEFT ? insets.left : 0),
    };

    return [style, paddingStyle];
  }, [style, edges, insets]);

  return <View style={appliedStyle} {...rest} />;
}
