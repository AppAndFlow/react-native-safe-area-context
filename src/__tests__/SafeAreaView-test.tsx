import { describe, expect, it } from '@jest/globals';
import { render } from '@testing-library/react-native';
import * as React from 'react';
import { View } from 'react-native';
import type { Metrics } from '../SafeArea.types';
import { SafeAreaProvider } from '../SafeAreaContext';
import { SafeAreaView } from '../SafeAreaView';

const INITIAL_METRICS: Metrics = {
  insets: { top: 1, left: 2, right: 3, bottom: 4 },
  frame: { x: 0, y: 0, height: 100, width: 100 },
};

describe('SafeAreaView', () => {
  it('renders', () => {
    const { toJSON } = render(
      <SafeAreaProvider initialMetrics={INITIAL_METRICS}>
        <SafeAreaView>
          <View />
        </SafeAreaView>
      </SafeAreaProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('can override padding styles', () => {
    const { toJSON } = render(
      <SafeAreaProvider initialMetrics={INITIAL_METRICS}>
        <SafeAreaView style={{ paddingTop: 0 }}>
          <View />
        </SafeAreaView>
      </SafeAreaProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('can set edges', () => {
    const { toJSON } = render(
      <SafeAreaProvider initialMetrics={INITIAL_METRICS}>
        <SafeAreaView edges={['top', 'bottom']}>
          <View />
        </SafeAreaView>
      </SafeAreaProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('can set edges value type', () => {
    const { toJSON } = render(
      <SafeAreaProvider initialMetrics={INITIAL_METRICS}>
        <SafeAreaView
          edges={{ top: 'additive', bottom: 'maximum' }}
          style={{ paddingTop: 24, paddingBottom: 24 }}
        >
          <View />
        </SafeAreaView>
      </SafeAreaProvider>,
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
