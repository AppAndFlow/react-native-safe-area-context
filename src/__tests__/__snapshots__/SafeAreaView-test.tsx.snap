// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SafeAreaView can override padding styles 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
    style={
      {
        "paddingTop": 0,
      }
    }
  >
    <View />
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;

exports[`SafeAreaView renders 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      [
        "bottom",
        "left",
        "right",
        "top",
      ]
    }
  >
    <View />
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
